bin/internal/prepare-environment.sh

```diff
@@ -0,0 +1,197 @@
+
+################################################################################
+# This script will prepare all the environment required to start a component.
+#
+# These environment variables should have already been loaded:
+# - INSTANCE_DIR
+#
+# Note: the INSTANCE_DIR can be predefined as global variable, or can be passed
+#       from command line "-c" parameter.
+#
+# Note: this script doesn't rely on <instance-dir>/bin/internal/run-zowe.sh, so
+#       it assumes all environment variables in `instance.env` are not loaded.
+#       With this assumption, we can safely call this script in Zowe Launcher
+#       which doesn't have the environment prepared by run-zowe.sh.
+#
+# Note: this script could be called multiple times during start up.
+################################################################################
+
+if [ "${ZWE_ENVIRONMENT_PREPARED}" = "true" ]; then
+  # environment variables are already prepared, skip running
+  return 0
+fi
+
+# export all variables defined in this script automatically
+set -a
+
+# initialize flag variable to avoid re-run this script
+ZWE_ENVIRONMENT_PREPARED=
+
+# if the user passes INSTANCE_DIR from command line parameter "-c"
+while getopts "c:" opt; do
+  case $opt in
+    c) INSTANCE_DIR=$OPTARG;;
+    \?)
+      echo "Invalid option: -$OPTARG" >&2
+      exit 1
+      ;;
+  esac
+done
+
+# validate INSTANCE_DIR which is required
+if [[ -z ${INSTANCE_DIR} ]]; then
+  echo "INSTANCE_DIR is not defined. You can either pass the value with -c parameter or define it as global environment variable." >&2
+  exit 1
+fi
+
+# find runtime directory
+ROOT_DIR=$(cd $(dirname $0)/../../;pwd)
+
+# read the instance environment variables to make sure they exists
+# Question: is there a better way to load these variables since this is already handled by
+#           <instance-dir>/bin/internal/run-zowe.sh
+. ${INSTANCE_DIR}/bin/internal/read-instance.sh
+if [ ! -z "${KEYSTORE_DIRECTORY}" -a -f "${KEYSTORE_DIRECTORY}/zowe-certificates.env" ]; then
+  . ${INSTANCE_DIR}/bin/internal/read-keystore.sh
+fi
+
+. ${ROOT_DIR}/bin/internal/zowe-set-env.sh
+
+# TODO think of a better name?
+WORKSPACE_DIR=${INSTANCE_DIR}/workspace
+
+# TODO - in for backwards compatibility, remove once naming conventions finalised and sorted #870
+ZOWE_APIM_GATEWAY_PORT=$GATEWAY_PORT
+ZOWE_IPADDRESS=$ZOWE_IP_ADDRESS
+ZOSMF_IP_ADDRESS=$ZOSMF_HOST
+VERIFY_CERTIFICATES=$ZOWE_APIM_VERIFY_CERTIFICATES
+ZOWE_NODE_HOME=$NODE_HOME
+ZOWE_JAVA_HOME=$JAVA_HOME
+
+# zip #1226 - 1.10 -> 1.9 backward compatibility - default keystore type if not supplied
+if [[ -z ${KEYSTORE_TYPE} ]]
+then
+  KEYSTORE_TYPE="PKCS12"
+fi
+
+LAUNCH_COMPONENTS=""
+export ZOWE_PREFIX=${ZOWE_PREFIX}${ZOWE_INSTANCE}
+ZOWE_DESKTOP=${ZOWE_PREFIX}DT
+
+# Make sure Java and Node are available on the Path
+. ${ROOT_DIR}/scripts/utils/configure-java.sh
+if [[ ${SKIP_NODE} != 1 ]]
+then
+  . ${ROOT_DIR}/scripts/utils/configure-node.sh
+fi
```

Comment:
Removed and will verify if this breaks anything.

Reply:
@stevenhorsman this change cause failure on some components. For example, I believe the proper way to fix it changing the components but I wonder if there are any extensions rely on the PATH. We can change this to be optional by checking the NODE_HOME JAVA_HOME value. If they are not empty, then we make sure the bin folders are in PATH. If the user doesn’t need Java or node, then they can leave those empty. What do you think?

Reply:
Good question. I'm making a change in those component lifecycle scripts right now, so I'll update them, but the bigger question of whether anyone other extenders are relying on them I'm not sure. You'd hope not and there is possibly a danger of an extender requiring a different version of node/java and us overriding, but I'm not sure how we cope with that anyway.
If we keep them in for compatibility then I'd like to switch it to use the `ensure_java_is_on_path` method from the utils/utils.sh

Reply:
Added `ensure_java_is_on_path` and `ensure_node_is_on_path` to `prepare-environment.sh`. I also changed those 2 functions to prepend the bin to PATH instead of append. This way I think we can make sure the java/node bin picked is the one we defined with JAVA_HOME and NODE_HOME.